cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -41,11 +41,127 @@ static int sim_ending = 0;
 
 #include "simulatormodule.h"
 #include <cocotb_utils.h>     // COCOTB_UNUSED
+#include <type_traits>
+
 
 struct module_state {
     PyObject *error;
 };
 
+
+/* define the extension types as templates */
+
+template<typename gpi_hdl>
+struct gpi_hdl_Object {
+    PyObject_HEAD
+    gpi_hdl hdl;
+
+    // The python type object, in a place that is easy to retrieve in templates
+    static PyTypeObject py_type;
+};
+
+template<typename gpi_hdl>
+static PyObject *gpi_hdl_repr(gpi_hdl_Object<gpi_hdl> *self) {
+    PyTypeObject *type = Py_TYPE(self);
+    return PyUnicode_FromFormat("<%s at %p>",
+                                  type->tp_name, self->hdl);
+}
+template<typename gpi_hdl>
+static Py_hash_t gpi_hdl_hash(gpi_hdl_Object<gpi_hdl> *self) {
+    Py_hash_t ret = (Py_hash_t)self->hdl;
+    // hash must never return -1
+    if (ret == (Py_hash_t)-1) {
+        ret = (Py_hash_t)-2;
+    }
+    return ret;
+}
+
+template<typename gpi_hdl>
+static PyObject *gpi_hdl_New(gpi_hdl hdl) {
+    if (hdl == NULL) {
+        Py_RETURN_NONE;
+    }
+    gpi_hdl_Object<gpi_hdl> *obj = PyObject_New(gpi_hdl_Object<gpi_hdl>, &gpi_hdl_Object<gpi_hdl>::py_type);
```

Comment:
Does `PyObject_New`'s implementation allow you to use `auto` here?

Reply:
Yes!